examples/multigpu_advtrain/evaluator.py

```diff
@@ -0,0 +1,333 @@
+    """
+    args = _FlagsWrapper(args or {})
+
+    assert args.batch_size, "Batch size was not given in args dict"
+    if X_test is None or Y_test is None:
+        raise ValueError("X_test argument and Y_test argument "
+                         "must be supplied.")
+    if model is None and predictions is None:
+        raise ValueError("One of model argument "
+                         "or predictions argument must be supplied.")
+    if model is not None:
+        warnings.warn("model argument is deprecated. "
+                      "Switch to predictions argument. "
+                      "model argument will be removed after 2018-01-05.")
+        if predictions is None:
+            predictions = model
+        else:
+            raise ValueError("Exactly one of model argument"
+                             " and predictions argument should be specified.")
+
+    # Define accuracy symbolically
+    correct_preds = tf.equal(tf.argmax(y, axis=-1),
+                             tf.argmax(predictions, axis=-1))
+
+    # acc_value = tf.reduce_mean(tf.to_float(correct_preds))
+
+    # Init result var
+    accuracy = 0.0
+
+    with sess.as_default():
+        # Compute number of batches
+        nb_batches = int(math.ceil(float(len(X_test)) / args.batch_size))
+        assert nb_batches * args.batch_size >= len(X_test)
+
+        X_cur = np.zeros((args.batch_size,) + X_test.shape[1:],
+                         dtype=X_test.dtype)
+        Y_cur = np.zeros((args.batch_size,) + Y_test.shape[1:],
+                         dtype=Y_test.dtype)
+        for batch in range(nb_batches):
+            # Must not use the `batch_indices` function here, because it
+            # repeats some examples.
+            # It's acceptable to repeat during training, but not eval.
+            start = batch * args.batch_size
+            end = min(len(X_test), start + args.batch_size)
+            cur_batch_size = end - start
+            X_cur[:cur_batch_size] = X_test[start:end]
+            Y_cur[:cur_batch_size] = Y_test[start:end]
+
+            # The last batch may be smaller than all others, so we need to
+            # account for variable batch size here
+            feed_dict = {x: X_cur, y: Y_cur}
+            if feed is not None:
+                feed_dict.update(feed)
+            # cur_acc = acc_value.eval(feed_dict=feed_dict)
+            cur_corr_preds = correct_preds.eval(feed_dict=feed_dict)
+
+            # accuracy += (cur_batch_size * cur_acc)
+            accuracy += cur_corr_preds[:cur_batch_size].sum()
+
+        assert end >= len(X_test)
+
+        # Divide by number of examples to get final value
+        accuracy /= len(X_test)
+
+    return accuracy
+
+
+def get_attack(model, x, attack_type, *args, **kwargs):
+    with tf.variable_scope(attack_type):
+        return get_attack_2(model, x, attack_type, *args, **kwargs)
+
+
+def get_attack_2(model, x, attack_type, sess, batch_size=None, y=None,
```

Comment:
It would probably be better to just provide a few config files / config scripts for each dataset, rather than to have this function with if statements that need to be maintained

Reply:
Simplified in 3312e05. There is now one small dictionary for all the shared parameters.